apps/communications/contacts/js/views/list.js

```diff
@@ -2160,6 +2245,130 @@ contacts.List = (function() {
+
+    var cache = [];
+
+    var groups = Array.prototype.slice.call(groupsList.children);
+    var contactsCount = CHUNK_SIZE;
+    while (contactsCount && groups.length) {
+      var group = groups.shift();
+
+      var groupCache = {};
+      groupCache.elementName = 'section';
+      groupCache.attributes = [];
+      for (var j = 0; j < group.attributes.length; j++) {
+        groupCache.attributes.push({
+          name: group.attributes[j].nodeName,
+          value: group.attributes[j].value
+        });
+      }
+
+      var section = document.createElement('section');
+      var header = group.querySelector('header');
+      // ICE group section has no header, only <ol> child
+      if (header) {
+        section.appendChild(header.cloneNode(true));
+      }
+
+      // We only want complete contacts until we reach CHUNK_SIZE
+      var ol = group.querySelector('ol');
+      var contacts = ol.cloneNode();
+      [].forEach.call(ol.querySelectorAll('li.contact-item'), (node) => {
+        if (!contactsCount) {
+          return;
+        }
+        contactsCount--;
+
+        node.dataset.cache = true;
+        var contact = node.cloneNode(true);
+
+        // We cannot cache the contact image, so we get rid of them.
+        var aside = contact.querySelector('aside');
+        if (aside) {
+          aside.parentNode.removeChild(aside);
+        }
+
+        contact.dataset.status = '';
+        contact.dataset.rendered = false;
+        contact.dataset.visited = false;
+        contacts.appendChild(contact);
+      });
+      section.appendChild(contacts);
+
+      groupCache.innerHTML = section.innerHTML;
+      section = null;
+      cache.push(groupCache);
+    }
+
+    return cache;
+  }
+
+  var cacheRequestTimer;
+  function cacheContactsList() {
+    // The Cache mechanism accesses localStorage.
+    // Use the following logic to cache the contact list in a
+    // moderate pattern.
+    // 1. When a request to cache the list is received, start a
+    //    timer.
+    // 2. If another request is received while the timer is
+    //    running, reset it.
+    // 3. Once the timer fires, start caching the list.
+
+    if (cacheRequestTimer) {
+      clearTimeout(cacheRequestTimer);
+    }
+
+    cacheRequestTimer = setTimeout(() => {
+      cacheRequestTimer = null;
+      Cache.firstChunk = getFirstChunkCache();
+      // Release memory taken by the old cache.
+      Cache.cleanup();
+    }, 1000);
+  }
+
+  function verifyAndRebuildCache() {
+    if (!Cache.active || !Cache.length) {
+      // Everything is fine.
+      cacheContactsList();
+      return;
+    }
+
+    // If there is still any contact to consume in the cache it
+    // means that this contact is no longer part of the contacts
+    // source, but it is still in the DOM, so we need to eliminate it.
+    for (var id of Cache.contacts) {
+      remove(id);
+      // We may also need to remove it from the list of ICE contacts.
+      ICEData.load().then(() => {
```

Comment:
I don't understand how we can get to the situation where we need to remove from the ice list when checking the cache.